project-schema.json: add lobbyingMeetings,Meetings,PublicOfficial obj…

odscjen opened the pull request:
…ects and minutes.lobbyingMeeting code

**Related issues**

closes #412 
closes #409 

**Description**

I used the `Address` object for `meeting.location` rather than `Location` as a meeting is going to happen at an address not a geometry or gazetteer entry.

**Merge checklist**

<!-- Complete the checklist before requesting a review. -->

- [x] [Log your changes](https://ocds-standard-development-handbook.readthedocs.io/en/latest/standard/contributing.html#logging-changes)

If there are changes to `project-schema.json` or `project-package-schema.json`:

- [x] Update the examples:
  - [ ] `docs/examples/example.json`
  - [ ] `docs/examples/blank.json`
- [x] Run `./manage.py pre-commit` to update `docs/_static/i8n.csv`

If you added a new definition to the schema, run `./manage.py pre-commit`.

If you added a new codelist:

  - [ ] Add an entry to `docs/reference/codelists.md`

odscjen committed:
project-schema.json: add lobbyingMeetings,Meetings,PublicOfficial objects and minutes.lobbyingMeeting code

odscjen committed:
manage.py pre-commit

odscjen committed:
update examples/

odscjen committed:
update changelog.md

odscjen committed:
fix build errors, change Meeting.location to ref Address

odscjen committed:
fix example errors

duncandewhurst commented on schema/project-level/project-schema.json:
```diff
@@ -1998,6 +2008,76 @@
+              "description": "The full name of the person.",
+              "type": "string"
+            },
+            "organization": {
+              "title": "Organization",
+              "description": "The organization in respect of which the person is a public official.",
+              "$ref": "#/definitions/OrganizationReference"
```
`.organization` should be a property of `PublicOfficial` rather than of `PublicOfficial.person` (`blank.json` and `example.json` will need updating too).

odscjen replied:
oops, I had misread that, fixed now

duncandewhurst commented on docs/reference/changelog.md:
```diff
@@ -28,6 +28,10 @@
   * `.transactions`
   * `.milestones`
   * `ContractingProcesses.milestones`
+* [#428](https://github.com/open-contracting/infrastructure/pull/428) - add objects:
+  * `Meeting`
+  * `lobbyingMeetings`
+  * `PublicOfficial`
```
`lobbyingMeetings` is a field rather than an object. I think it's fine to only mention the field since the `Meeting` and `PublicOfficial` definitions are only used in that context.

```suggestion
* [#428](https://github.com/open-contracting/infrastructure/pull/428) - add `lobbyingMeetings` field.
```

duncandewhurst commented on schema/project-level/project-schema.json:
```diff
@@ -1998,6 +2008,76 @@
+          ],
+          "format": "date-time"
+        },
+        "location": {
+          "title": "Location",
+          "description": "The location of the meeting.",
```
Update the name, title and description to match the referenced definition.

```suggestion
        "address": {
          "title": "Address",
          "description": "The address of the meeting.",
```

duncandewhurst reviewed (changes requested):
Suggest addressing the remainder of https://github.com/open-contracting/infrastructure/issues/409 in this PR too.

Please check and update the mapping and example for the related CoST IDS elements in `sustainability.yaml`

odscjen committed:
apply review suggestions

odscjen committed:
add social.consultationMeetings and update sustainability.yaml

odscjen committed:
update examples

odscjen committed:
fix typo

odscjen committed:
fix examples

odscjen commented:
I've added the rest of #409, I've held off on that as it meant creating the parent `Social` object and I was waiting until all the related issues for that were agreed but I suppose it does make sense to do it as part of this. Also sorry for all the commits, I updated Chrome to the latest version but they've not realised a ChromeDriver version for it yet so I can't run the build tests on my local machine.

duncandewhurst committed:
Merge branch '0.9-dev' into 412_lobbying_transparency

duncandewhurst committed:
docs/examples/example.json: Remove consultationMeetings.publicOfficial

duncandewhurst committed:
mapping/sustainability.yaml: Link public consultation meetings

duncandewhurst committed:
docs/cost/ids/sustainability.md: Remove empty examples

duncandewhurst commented on mapping/sustainability.yaml:
```diff
@@ -1209,16 +1209,16 @@
         For each meeting:
 
-        1. Publish the meeting invite. Add a document, set its `.documentType` to 'consultationMeetingInvite' and its `.url` to the URL at which the meeting invite is available.
+        1. Publish the meeting invite. Add a document, set its `.documentType` to 'consultationMeetingInvitation' and its `.url` to the URL at which the meeting invite is available.
```
Refer to the common operation for adding a project document.

```suggestion
        1. Publish the meeting invite. [Add a project document](../common.md#add-a-project-document) and set its `.documentType` to 'consultationMeetingInvitation'.
```

duncandewhurst commented on mapping/sustainability.yaml:
```diff
@@ -1209,16 +1209,16 @@
+        - `.numberOfparticipants` to the number of people that participated in the meeting
 
-        3. Publish the meeting minutes. Add a document, set its `.documentType` to 'consultationMeetingMinutes' and its `.url` to the URL at which the meeting minutes are available.
+        3. Publish the meeting minutes. Add a document, set its `.documentType` to 'minutes.consultationMeeting' and its `.url` to the URL at which the meeting minutes are available.
```
Refer to the common operation for adding a project document.

```suggestion
        3. Publish the meeting minutes. [Add a project document](../common.md#add-a-project-document), set its `.documentType` to 'minutes.consultationMeeting'.
```

duncandewhurst commented on mapping/sustainability.yaml:
```diff
@@ -1714,16 +1714,17 @@
-        1. Publish the meeting agenda. Add a document, set its `.documentType` to 'lobbyingMeetingAgenda' and its `.url` to the URL at which the agenda is available.
-
-        2. Publish the meeting details. Add a `Meeting` object to the `.lobbyingMeetings` array and set:
+        1. Publish the meeting details. Add a `Meeting` object to the `.lobbyingMeetings` array and set:
```
Explain what to do if the public official's organization doesn't exist yet. Copy-edit to avoid referencing the public official's organization before it is introduced (GitHub won't let me add a multi-line comment today so you'll also need to delete the subsequent lines).

```suggestion
        1. Publish the meeting details:
          1. Add a `Meeting` object to the `.lobbyingMeetings` array and set:
            - `.id` incrementally
            - `.date` to the date of the meeting
            - `.address` to the address of the meeting
            - `.numberOfParticipants` to the number of people present at the meeting
            - `.publicOfficial.name` to the name of the public official present at the meeting
            - `.publicOfficial.jobTitle` to the job title of the public official present at the meeting
          1. Get the `Organization` in `.parties` that represents the public official's organization. If none exists yet, [add an organization](../common.md#add-an-organization).
          1. Set the meeting's `.publicOfficial.organization` to the `.id` and `.name` of the organization.  
```

duncandewhurst commented on mapping/sustainability.yaml:
```diff
@@ -1714,16 +1714,17 @@
+          ` `.publicOfficial.jobTitle`to the job title of the person named in`.name\`
 
-        3. Publish the meeting minutes. Add a document, set its `.documentType` to 'lobbyingMeetingMinutes' and its `.url` to the URL at which the meeting minutes are available.
+        2. Publish the meeting minutes. Add a document, set its `.documentType` to 'minutes.lobbyingMeeting' and its `.url` to the URL at which the meeting minutes are available.
```
Reference the common operation for adding a project document:

```suggestion
        2. Publish the meeting minutes. [Add a project document](../common.md#add-a-project-document) and set its `.documentType` to 'minutes.lobbyingMeeting'.
```

duncandewhurst reviewed (changes requested):
Happy for you to request a review from James once you've resolved my comments.

I also added a few commits to remove `consultationMeetings.publicOfficial` from `example.json`, to link the public consultation meetings element in the climate finance module to the public consultation meetings element in the social module, and to remove empty examples from `sustainability.md`. I can respond to any comments from James on those changes.

odscjen committed:
Apply suggestions from code review

Co-authored-by: Duncan Dewhurst <duncan.dewhurst@opendataservices.coop>

odscjen committed:
Merge branch '0.9-dev' into 412_lobbying_transparency

duncandewhurst commented:
@jpmckinney if you could prioritise this PR review above the other OC4IDS PR reviews that would be great, as it contains some changes to make it possible to link to individual CoST IDS elements in `sustainability.md`, which it would be helpful to make use of in other PRs.

jpmckinney commented on schema/project-level/project-schema.json:
```diff
@@ -2336,6 +2351,92 @@
+          "description": "The number of people in attendance at the meeting.",
+          "type": "integer"
+        },
+        "publicOfficial": {
```
And corresponding changes elsewhere. The "official" is always the person, so we need to refer to this as the "office" (i.e. the position), otherwise it's confusing.

```suggestion
        "publicOffice": {
```

jpmckinney commented on manage.py:
````diff
@@ -942,9 +945,7 @@ def update_sustainability_docs():
-              "\n```json\n",
-              element.get("example", ""),
-              "\n```",
+              f"\n```json\n{element.get('example', '')}\n```" if element.get('example', '') != '' else '',
````
```suggestion
              f"\n```json\n{element['example']}\n```" if element.get('example') else '',
```

jpmckinney commented on schema/project-level/codelists/documentType.csv:
```diff
@@ -64,3 +64,6 @@ contractAmendment,Contract amendment,A legal document that changes or adds to th
 escalationApproval,Escalation approval,Documentation of one or more approved escalations of the contract price during the project life-cycle.,OC4IDS
+minutes.lobbyingMeeting,Lobbying meeting minutes,"A record of the issues discussed, agreements reached and decisions taken at a lobbying meeting.",OC4IDS
+minutes.consultationMeeting,Consultation meeting minutes,"A record of the issues discussed, agreements reached and decisions taken at a public consultation meeting.",OC4IDS
+consultationMeetingInvitation,Consultation meeting invitation,A document detailing the occurrence of a public consultation meeting relating to the project.,OC4IDS
```
The description didn't really reflect that the invitation occurs before the meeting.

```suggestion
consultationMeetingInvitation,Consultation meeting invitation,A notice about an upcoming public consultation meeting relating to the project.,OC4IDS
```

jpmckinney commented on schema/project-level/project-schema.json:
```diff
@@ -297,6 +297,21 @@
     },
+    "lobbyingMeetings": {
+      "title": "Lobbying meetings",
+      "description": "Information about lobbying meetings in relation to the project. That is, meetings with public officials that are made, managed or directed with the purpose of influencing public decision-making in relation to the project, whether for private, public or collective ends, whether for compensation or without. Public decision-making is the creation and amendment of legislation or ...
```
```suggestion
      "description": "Information about lobbying meetings in relation to the project. That is, meetings with public officials that are made, managed or directed with the purpose of influencing public decision-making in relation to the project, whether for private, public or collective ends, whether for compensation or without. Public decision-making is the creation and amendment of legislation or any other regulatory measures; the development, modification and implementation of public policies, strategies and programs; and the awarding of government contracts or grants, administrative decisions or any other public spending decisions.",
```

jpmckinney commented on schema/project-level/project-schema.json:
```diff
@@ -297,6 +297,21 @@
+    },
+    "social": {
+      "title": "Social",
+      "description": "Information about the social impact of the project and actions the project will take to mitigate potentially negative impacts on adjacent populations and project workers.",
```
To align with changes to `environment` in another PR.

Repeat the changes for the Social definition.


```suggestion
      "description": "Information about the social impacts of the project and the actions the project will take to mitigate potentially negative impacts on adjacent populations and project workers.",
```

jpmckinney commented on schema/project-level/project-schema.json:
```diff
@@ -2336,6 +2351,92 @@
+    },
+    "Meeting": {
+      "title": "Meeting",
+      "description": "The details of a meeting.",
```
```suggestion
      "description": "Information about a meeting.",
```

jpmckinney commented on schema/project-level/project-schema.json:
```diff
@@ -2336,6 +2351,92 @@
+      "properties": {
+        "id": {
+          "title": "ID",
+          "description": "A locally unique identifier for the meeting.",
```
```suggestion
          "description": "The locally unique identifier for the meeting.",
```

jpmckinney commented on schema/project-level/project-schema.json:
```diff
@@ -2336,6 +2351,92 @@
+          "$ref": "#/definitions/Address"
+        },
+        "numberOfParticipants": {
+          "title": "Participant count",
```
```suggestion
          "title": "Number of participants",
```

jpmckinney commented on schema/project-level/project-schema.json:
```diff
@@ -2336,6 +2351,92 @@
+      "properties": {
+        "person": {
+          "title": "Person.",
+          "description": "The person.",
```
```suggestion
          "description": "The person holding the public office.",
```

jpmckinney commented on schema/project-level/project-schema.json:
```diff
@@ -2336,6 +2351,92 @@
+        },
+        "jobTitle": {
+          "title": "Job title",
+          "description": "The job title of the person with respect to the organization.",
```
```suggestion
          "description": "The job title of the person holding the public office.",
```

jpmckinney commented on schema/project-level/project-schema.json:
```diff
@@ -2336,6 +2351,92 @@
+        },
+        "organization": {
+          "title": "Organization",
+          "description": "The organization in respect of which the person is a public official.",
```
```suggestion
          "description": "The organization in which the public office exists.",
```

duncandewhurst committed:
Update manage.py

Co-authored-by: James McKinney <26463+jpmckinney@users.noreply.github.com>

odscjen committed:
Apply suggestions from code review

Co-authored-by: James McKinney <26463+jpmckinney@users.noreply.github.com>

odscjen committed:
apply rest of review suggestions and remove duplications in sustainability.yaml

odscjen committed:
Merge branch '0.9-dev' into 412_lobbying_transparency

odscjen commented:
@jpmckinney all suggestions now applied, are you ok for this to be merged or do you need to check again?

jpmckinney commented on manage.py:
```diff
@@ -338,7 +339,8 @@ def update_sub_schema_reference(schema):
                          'parties/0/people/0/address',
-                         'parties/0/people/0/identifier']
+                         'parties/0/people/0/identifier',
+                         'social/consultationMeetings/0/publicOfficial']
```
```suggestion
                         'social/consultationMeetings/0/publicOffice']
```

jpmckinney commented on docs/examples/blank.json:
```diff
@@ -476,6 +476,56 @@
+      },
+      "date": "date-time",
+      "numberOfParticipants": "integer",
+      "PublicOffice": {
```
```suggestion
      "publicOffice": {
```

jpmckinney commented on schema/project-level/project-schema.json:
```diff
@@ -2370,6 +2385,92 @@
+      "type": "object",
+      "properties": {
+        "person": {
+          "title": "Person.",
```
```suggestion
          "title": "Person",
```

jpmckinney reviewed (commented):
Just some very small edits, then can be merged.

duncandewhurst committed:
Apply suggestions from code review

Co-authored-by: James McKinney <26463+jpmckinney@users.noreply.github.com>

duncandewhurst commented:
Given that there are a few other PRs that depend on this one, I've committed the suggestions and will merge it, just to speed things along :-)
